pkg/cmd/metrics-loader/watcher.go

```diff
@@ -0,0 +1,153 @@
+// provided Processor.
+func NewGoogleWatcher(
+	ctx context.Context, processor Processor, config GoogleWatcherConfig,
+) (*GoogleWatcher, error) {
+	storage, err := NewGoogleStorage()
+	if err != nil {
+		return nil, err
+	}
+	return &GoogleWatcher{
+		GoogleWatcherConfig: config,
+		processor:           processor,
+		googleStorage:       storage,
+		ctx:                 ctx,
+	}, nil
+}
+
+func (w *GoogleWatcher) Start() error {
+	log.Print("Starting watcher")
+	// Set up Google Cloud Pub/Sub client.
+	pubsubClient, err := pubsub.NewClient(w.ctx, w.ProjectID)
+	if err != nil {
+		return errors.Wrap(err, "failed to create pubsub client")
+	}
+
+	// Delete Google Cloud Pub/Sub topic if it already exists.
+	topic := pubsubClient.Topic(w.TopicID)
+	exists, err := topic.Exists(w.ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if pubsub topic exists")
+	}
+	if !exists {
+		// Set up Google Cloud Pub/Sub topic.
+		topic, err = pubsubClient.CreateTopic(w.ctx, w.TopicID)
+		if err != nil {
+			return errors.Wrap(err, "failed to create pubsub topic")
+		}
+	}
+
+	// Set up Google Cloud Storage bucket notification.
+	err = w.googleStorage.RegisterFinalizeNotification(w.BucketName, w.ProjectID, w.TopicID)
+	if err != nil {
+		return errors.Wrap(err, "failed to register bucket notification")
+	}
+
+	sub := pubsubClient.Subscription(w.SubscriptionID)
+	exists, err = sub.Exists(w.ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if pubsub subscription exists")
+	}
+	if !exists {
+		// Subscribe to Google Cloud Pub/Sub subscription.
+		sub, err = pubsubClient.CreateSubscription(w.ctx, w.SubscriptionID, pubsub.SubscriptionConfig{
+			Topic: topic,
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to create pubsub subscription")
+		}
+	}
+	log.Printf("Watcher started for subscription %q", sub.String())
+
+	return sub.Receive(w.ctx, func(ctx context.Context, msg *pubsub.Message) {
+		defer msg.Ack()
+
+		var data StorageData
+		if err := json.Unmarshal(msg.Data, &data); err != nil {
+			log.Printf("Failed to unmarshal event data: %v", err)
+			return
+		}
+
+		// Only process files in the configured bucket path with the `.om`
+		// extension.
+		dir := path.Dir(data.Name)
+		if dir != w.BucketWatchPath {
+			return
+		}
+		if !strings.HasSuffix(data.Name, ".om") {
+			return
+		}
+
+		log.Printf("Received bucket file event %q", data.Name)
+		uri := fmt.Sprintf("gs://%s/%s", data.Bucket, data.Name)
+		if err := w.processFile(uri); err != nil {
+			log.Printf("Failed to process file %q: %v", uri, err)
+		}
+		log.Printf("Finished processing file %q", uri)
+
+	})
+}
+
+func (w *GoogleWatcher) processFile(uri string) error {
+	reader, err := w.googleStorage.Read(uri)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	err = w.processor.Process(reader)
+	if err != nil {
+		return err
+	}
+	return w.googleStorage.Move(uri, w.toProcessedPath(uri))
```

Comment:
If `Process` already sends the metrics to prom, whats the reason for keeping the file around as opposed to deleting it?